### ISSUE TITLE
Wrong results for GAPW with mGGA functionals

### ISSUE THREAD

**Issue by mkrack** [cp2k/cp2k#1116]
Energy and potential are not consistent using GAPW with meta-GGA functionals. An OT CG SCF run for a water molecule using GAPW with TPSS does not converge. A debug run with diagonalization shows wrong analytical (atomic) forces as discussed [here](https://groups.google.com/g/cp2k/c/K4A-gWzZrMc/m/wRWjktVIBgAl)

**Comment by abussy** [comment]
I am somewhat familiar with GAPW, I'll have a look.

**Comment by mkrack** [comment]
The following H2O input still does not converge with GAPW for the mGGA functional TPSS using OT CG indicating that the issue has not been resolved fully. OT DIIS converges with the basis set DZVP-ALL instead of 6-31Gxx, but there is then still a force mismatch detected.
@abussy Please, could you check again.
```
@SET XCFUN     TPSS

@SET BASIS_SET 6-31Gxx
@SET POTENTIAL ALL

@SET GAPW      0
@IF ${POTENTIAL} == ALL
@SET GAPW      1
@ENDIF

&GLOBAL
 PROJECT H2O_${XCFUN}
 RUN_TYPE debug
 PRINT_LEVEL medium
 PREFERRED_DIAG_LIBRARY sl
&END GLOBAL

&DEBUG
 DEBUG_FORCES yes
 DEBUG_STRESS_TENSOR no
&END DEBUG

&FORCE_EVAL
 METHOD Quickstep
 &DFT
  BASIS_SET_FILE_NAME BASIS_SET
  BASIS_SET_FILE_NAME BASIS_MOLOPT
  BASIS_SET_FILE_NAME EMSL_BASIS_SETS
  BASIS_SET_FILE_NAME GTH_BASIS_SETS
  POTENTIAL_FILE_NAME POTENTIAL
  CHARGE 0
  MULTIPLICITY 1
  &MGRID
   CUTOFF 800
   NGRIDS 5
   REL_CUTOFF 60
  &END MGRID
  &QS
   EPS_DEFAULT 1.0E-14
@IF ${GAPW}
   METHOD GAPW
   ALPHA0_H 10.0
   EPSFIT 1.0E-8
   EPSISO 1.0E-12
   EPSSVD 0.0
   EPSRHO0 1.0E-8
   LMAXN0 4
   LMAXN1 6
   QUADRATURE GC_LOG
@ENDIF
  &END QS
  &SCF
   EPS_SCF 3.0E-7
   MAX_SCF 21
   SCF_GUESS atomic
   &OT on
    MINIMIZER CG
    PRECONDITIONER FULL_ALL
   &END OT
   &OUTER_SCF on
    EPS_SCF 3.0E-7
    MAX_SCF 30
   &END OUTER_SCF
   &PRINT
    &RESTART
     BACKUP_COPIES 0
    &END RESTART
   &END PRINT
  &END SCF
  &XC
   &XC_FUNCTIONAL ${XCFUN}
   &END XC_FUNCTIONAL
  &END XC
 &END DFT
 &SUBSYS
  &CELL
   ABC 10.0 10.0 10.0
  &END CELL
  &COORD
   O    0.0000   0.0000   0.1180
   H    0.0000   0.7602  -0.4710
   H    0.0000  -0.7602  -0.4710
  &END COORD
  &KIND H
   ELEMENT H
   BASIS_SET ${BASIS_SET}
   POTENTIAL ${POTENTIAL}
   LEBEDEV_GRID 110
   RADIAL_GRID 200
  &END KIND
  &KIND O
   ELEMENT O
   BASIS_SET ${BASIS_SET}
   POTENTIAL ${POTENTIAL}
   LEBEDEV_GRID 110
   RADIAL_GRID 200
  &END KIND
 &END SUBSYS
&END FORCE_EVAL
```

**Comment by abussy** [comment]
@mkrack Thanks for the feedback. This problem seems to only appear with OT. When I run your input with standard diagonalization it works fine. I double checked my math and implementation and couldn't find a mistake (for now at least). As of now, I am quite puzzled as to why OT fails.

**Comment by mkrack** [comment]
A convergence failure of OT CG indicates an inconsistency of (KS) potential and energy, since the wavefunction gradient combined with just energy calculation in the LS step is employed in the SCF iteration cycle. Diagonalisation and OT DIIS are not sensible in that way and might converge without problem. Thus a failure of OT CG is a first indicator that something is fishy, i.e. either energy or potential are wrong. Switching to GPW while keeping TPSS as XCFUN or to PBE while keeping GAPW as method in the test input allows for convergence with OT CG. I think you fixed one bug in GAPW, but there might be other issues with GAPW for mGGAs. I observe also a dependence on the basis set. For instance the DZVP-ALL basis set behaves less problematic than the 6-31G** basis set.

**Comment by abussy** [comment]
Looking at the GPW code in `grid_integrate.F`:
https://github.com/cp2k/cp2k/blob/d7391792f6eb30a7d8110956cb79f1d2a46025b9/src/grid/grid_integrate.F#L399-L484

and comparing to the GAPW forces in `qs_ks_atom.F`, it seems that the whole `IF` clause regarding tau functionals is completely ignored/forgotten. This is something that I oversaw as I was hunting for tau code to correct and this did not show up.

There is definitely more work to be done and I guess that I was "lucky" in my previous tests that the tau influence on forces was small. I do not think this will fix OT convergence issues though.

**Comment by oschuett** [comment]
@abussy, I'm just about to migrate that code from `grid_integrate.F` to C (#1144). In the process I discovered a very compact notation. So, maybe it's actually easier to fix the problem in the new code?

**Comment by abussy** [comment]
@oschuett `grid_integrate.F` and the general treatment of meta-GGA functionals in GPW are fine. The problem lies in `qs_ks_atom.F`, so most likely the C migration will not affect this issue.

**Comment by abussy** [comment]
I did some more work and came up with this PR  #1169. @mkrack with the new code, your input for TPSS water runs fine, for both  DZVP-ALL and 6-31Gxx basis sets, using the CG minimizer for OT.

I also speculated that atomic forces were wrong due to some explicit treatment of tau functionals in GPW ( `grid_integrate.F`), which was absent from GAPW. Upon closer inspection, I realised that it is implicitly done in `qs_ks_atom.F`. Successful DEBUG runs for the forces seem to suggest that the atomic force treatment in metaGGA GAPW is actually correct.
